### PR TITLE
Resolves #6 implement asm translator with documentation

### LINKED ISSUE #6: Implement ASM translator in Python
We need it as a part of example for simple implementation.

We need to implement very simple asm language for brainfuck ISA. Isa description we already have. It should be an implementation with a lot of documentation.

Also:

- it should be implemented in separated module: `python/translator_asm.py`;
- we need to describe it in `./README.md` and in `./python/README.md` (add section `Язык программирования Asm` and other mention in the file);
- we need to document asm language for report, but it should be placed not in general README.md file

### PR DESCRIPTION
**Pull request by k1b24**
Реализация транслятора asm 

Лазеев Сергей P33131

### CODE REVIEW

File: python/README_asm.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+instr ::= "increment"
+        | "decrement"
+        | "right"
+        | "left"
+        | "print"
+        | "input"
+        | "jmp" arg
+        | "jz" arg
+
+arg ::= integer
```
Comment (ryukzak) [inline python/README_asm.md:25]: Do we really need `integer` here? label_name should be enough.

File: python/README_asm.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+label ::= label_name ":"
+
+instr ::= "increment"
+        | "decrement"
+        | "right"
+        | "left"
+        | "print"
+        | "input"
+        | "jmp" arg
+        | "jz" arg
```
Comment (ryukzak) [inline python/README_asm.md:23]:
```suggestion
instr ::= op0 
        | op1 arg
        
op0 ::= "increment" 
      | "decrement"
      | "right" 
      | "left"
      | "print" 
      | "input"
      
op1 ::= "jmp"
      | "jz"
```

File: python/README_asm.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+Синтаксис в расширенной БНФ.
+
+- `[ ... ]` -- вхождение 0 или 1 раз
+- `{ ... }` -- повторение 0 или несколько раз
+- `{ ... }-` -- повторение 1 или несколько раз
+
+``` ebnf
+program ::= { line }
+
+line ::= [ label ] [ instr ] [ comment ] "\n"
```
Comment (ryukzak) [inline python/README_asm.md:12]: To be more simple:

```suggestion
line ::= label: [ comment ] "\n"
       | instr [ comment ] "\n"
       | [ comment ] "\n"
```

File: python/README_asm.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+Пример кода на Asm см. в файле [./examples/cat.asm](./examples/cat.asm).
+
+## Транслятор
+
+Интерфейс командной строки: `translator_asm.py <input_file> <target_file>`
+
+Реализован в модуле [translator_asm](./translator_asm.py)
+
+Ассемблерные мнемоники один в один транслируются в машинные команды.
+Метки из программы исчезают, а на место обращений к ним подставляются конкретные адреса.
```
Comment (ryukzak) [inline python/README_asm.md:73]: We need golden tests for the translator.

Comment (ryukzak) [inline python/README_asm.md:73]: Two paragraphs?

File: python/translator_asm.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            instr["arg"] = labels[label]
+
+    return instrs
+
+
+def parse(text):
+    """Разбор текста на словарь с определением меток и список машинных инструкций."""
+    pc = 0
+    labels = {}
+    instrs = []
```
Comment (ryukzak) [inline python/translator_asm.py:37]: Rename to `code`, as in `translator.py`

**Review by ryukzak — CHANGES_REQUESTED** [review]
In general -- looks well. But as education purpose I plan to significantly rewrite your code to be more simple after you finish these fixes.

I will call you for review my PR.

**Comment by k1b24** [comment]
@ryukzak may you take a look, please

**Review by ryukzak — CHANGES_REQUESTED** [review]
Сделал вам ПР с рефакторингом: https://github.com/k1b24/brainfuck/pull/1

Можете если необходимо документацию обновить.

File: python/integration_test.py [hunk 6]
```diff
@@ -60,8 +62,12 @@ def test_translator_and_machine(golden, caplog):
 
     # Создаём временную папку для тестирования приложения.
     with tempfile.TemporaryDirectory() as tmpdirname:
+        # Определяем язык golden теста. По умолчанию предполагается язык Brainfuck
+        lang = golden.get("lang", "bf")
+        assert lang == "bf" or lang == "asm"
```
[the remark this answers is not in the document]
Comment (k1b24) [reply python/integration_test.py:67]: It seems like this code will be a little bigger, if I get you right

`
lang = None
source = None
if "in_source_asm" in golden:
  lang = "asm"
  source = golden["in_source_asm"]
if "in_source" in golden:
  assert lang is None, "both in_source and in_source_asm are present"
  lang = "bf"
  source = golden["in_source"]
assert lang is not None, "either in_source xor in_source_asm must be set"
`

And in this case we also need to add some logic about referencing the `in_source` param on this line below:
`file.write(golden["in_source"])`

**Comment by k1b24** [comment]
Merged PR and fixed some typos in the documentation

File: python/integration_test.py (hunk 6 above)
Comment (ryukzak) [inline python/integration_test.py:67]:
```suggestion
        assert "in_source" in golden !=  "in_source_asm" in golden
```

File: python/integration_test.py [hunk 7]
```diff
@@ -60,8 +62,12 @@ def test_translator_and_machine(golden, caplog):
...
+
         # Готовим имена файлов для входных и выходных данных.
-        source = os.path.join(tmpdirname, "source.bf")
+        source = os.path.join(tmpdirname, "source." + lang)
```
Comment (ryukzak) [inline python/integration_test.py:70]:
```suggestion
        source = os.path.join(tmpdirname, "source")
```

No body see this files...

File: python/integration_test.py [hunk 8]
```diff
@@ -74,7 +80,10 @@ def test_translator_and_machine(golden, caplog):
         # Запускаем транслятор и собираем весь стандартный вывод в переменную
         # stdout
         with contextlib.redirect_stdout(io.StringIO()) as stdout:
-            translator.main(source, target)
+            if lang == "bf":
+                translator.main(source, target)
+            elif lang == "asm":
+                translator_asm.main(source, target)
```
Comment (ryukzak) [inline python/integration_test.py:86]:
```suggestion
            if "in_source" in golden:
                with open(source, "w", encoding="utf-8") as file:
                    file.write(golden["in_source"])
                translator.main(source, target)
            elif ...
```

File: python/integration_test.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ryukzak) [reply python/integration_test.py:67]: Just an idea

**Review by ryukzak — CHANGES_REQUESTED** [review]

**Pull request merged**